src/lib/mongoBackend/mongoSubCache.cpp

```diff
@@ -0,0 +1,1204 @@
+  //
+  for (unsigned int ix = 0; ix < condVec.size(); ++ix)
+  {
+    BSONObj                   condition = condVec[ix].embeddedObject();
+    std::string               condType;
+    std::vector<BSONElement>  valueVec;
+
+    condType = condition.getStringField(CSUB_CONDITIONS_TYPE);
+    if (condType != "ONCHANGE")
+    {
+      continue;
+    }
+
+    NotifyCondition* ncP = new NotifyCondition();
+    ncP->type = condType;
+
+    valueVec = condition.getField(CSUB_CONDITIONS_VALUE).Array();
+    for (unsigned int vIx = 0; vIx < valueVec.size(); ++vIx)
+    {
+      std::string condValue;
+
+      condValue = valueVec[vIx].String();
+      ncP->condValueList.push_back(condValue);
+    }
+
+    cSubP->notifyConditionVector.push_back(ncP);
+  }
+
+  if (cSubP->notifyConditionVector.size() == 0)  // Cleanup
+  {
+    LM_E(("ERROR (empty notifyConditionVector) - cleaning up"));
+    cachedSubscriptionDestroy(cSubP);
+    delete cSubP;
+    return -4;
+  }
+
+  mongoSubCacheItemInsert(cSubP);
+
+  return 0;
+}
+
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheItemInsert - create a new sub, fill it in, and add it to cache
+*/
+void mongoSubCacheItemInsert
+(
+  const char*               tenant,
+  const char*               servicePath,
+  SubscribeContextRequest*  scrP,
+  const char*               subscriptionId,
+  int64_t                   expirationTime,
+  int64_t                   throttling,
+  Format                    notifyFormat
+)
+{
+  CachedSubscription* cSubP = new CachedSubscription();
+  LM_T(LmtMongoSubCache,  ("allocated CachedSubscription at %p", cSubP));
+
+  //
+  // 1. First the non-complex values
+  //
+  cSubP->tenant                = (tenant[0] == 0)? NULL : strdup(tenant);
+  cSubP->servicePath           = strdup(servicePath);
+  cSubP->subscriptionId        = strdup(subscriptionId);
+  cSubP->reference             = strdup(scrP->reference.get().c_str());
+  cSubP->expirationTime        = expirationTime;
+  cSubP->throttling            = throttling;
+  cSubP->lastNotificationTime  = 0;
+  cSubP->pendingNotifications  = 0;
+  cSubP->notifyFormat          = notifyFormat;
+  cSubP->next                  = NULL;
+
+  LM_T(LmtMongoSubCache, ("inserting a new sub in cache (%s). lastNotifictionTime: %lu", cSubP->subscriptionId, cSubP->lastNotificationTime));
+
+
+  //
+  // 2. Then the values that have functions/methods for filling/parsing
+  //
+  cSubP->throttling = scrP->throttling.parse();
+  cSubP->notifyConditionVector.fill(scrP->notifyConditionVector);
+
+
+  //
+  // 3. Convert all patterned EntityId to EntityInfo
+  //
+  for (unsigned int ix = 0; ix < scrP->entityIdVector.vec.size(); ++ix)
+  {
+    EntityId* eIdP;
+
+    eIdP = scrP->entityIdVector.vec[ix];
+
+    if (eIdP->isPattern == "false")
+    {
+      continue;
+    }
+
+    EntityInfo* eP = new EntityInfo(eIdP->id, eIdP->type);
+    
+    cSubP->entityIdInfos.push_back(eP);
+  }
+
```

Comment:
This condition is taken care of before calling the function (this function is only called in one place - in mongoSubscribeContext;

    //
    // Add the subscription to the subscription cache.
    // But only if any of the entities in entityIdVector is pattern based -
    // AND it is a subscription of type ONCHANGE
    //
    bool patternBased = false;
    bool onchange     = false;

    // 1. Pattern-based?
    for (unsigned int ix = 0; ix < requestP->entityIdVector.size(); ++ix)
    {
      if (requestP->entityIdVector[ix]->isPattern == "true")
      {
       	patternBased = true;
	break;
      }
    }

    // 2. ONCHANGE?
    for (unsigned int ix = 0; ix < requestP->notifyConditionVector.size(); ++ix)
    {
      if (strcasecmp(requestP->notifyConditionVector[ix]->type.c_str(), ON_CHANGE_CONDITION) == 0)
      {
        onchange = true;
        break;
      }
    }

    // 3. Create Subscription for the cache
    if (patternBased && onchange)
    {
      std::string oidString = oid.toString();

      mongoSubCacheItemInsert(tenant.c_str(), servicePath.c_str(), requestP, oidString.c_str(), expiration, throttling, notifyFormat);
    }

Other insert-functions implement that check, whereas in this function is is not necessary (OK, this could be argued with "what if somebody goes and uses this function from some other place?" And it would be true, of course.  At the same time, having the same check in each layer doesn't make any sense it is just a waste of time. I'll add a mention of this as a prerequisite in the function-header.)